lib/Frontend/ArgsToFrontendOutputsConverter.cpp

```diff
@@ -0,0 +1,182 @@
+    const ArgList &args, DiagnosticEngine &diags) {
+  if (const Arg *A = args.getLastArg(options::OPT_output_filelist)) {
+    assert(!args.hasArg(options::OPT_o) &&
+           "don't use -o with -output-filelist");
+    return ArgsToFrontendOutputsConverter::readOutputFileList(A->getValue(),
+                                                              diags);
+  }
+  return args.getAllArgValues(options::OPT_o);
+}
+
+OutputFilesComputer::OutputFilesComputer(
+    const ArgList &args, DiagnosticEngine &diags,
+    const FrontendInputsAndOutputs &inputsAndOutputs)
+    : Args(args), Diags(diags), InputsAndOutputs(inputsAndOutputs),
+      OutputFileArguments(
+          getOutputFilenamesFromCommandLineOrFilelist(Args, Diags)),
+      OutputDirectoryArgument(
+          OutputFileArguments && OutputFileArguments->size() == 1 &&
+                  llvm::sys::fs::is_directory(OutputFileArguments->front())
+              ? StringRef(OutputFileArguments->front())
+              : StringRef()),
+      DoOutputFileArgumentsMatchInputs(
+          OutputDirectoryArgument.empty() && OutputFileArguments &&
+          OutputFileArguments->size() ==
+              InputsAndOutputs.countOfInputsProducingOutput()),
+      FirstInput(InputsAndOutputs.hasSingleInput()
+                     ? InputsAndOutputs.getFilenameOfFirstInput()
+                     : StringRef()),
+      RequestedAction(
+          ArgsToFrontendOptionsConverter::determineRequestedAction(Args)),
+      ModuleNameArg(Args.getLastArg(options::OPT_module_name)),
+      Suffix(FrontendOptions::suffixForPrincipalOutputFileForAction(
+          RequestedAction)),
+      HasTextualOutput(
+          FrontendOptions::doesActionProduceTextualOutput(RequestedAction)) {}
+
+Optional<std::vector<std::string>>
+OutputFilesComputer::computeOutputFiles() const {
+  if (!OutputFileArguments)
+    return None;
+
+  std::vector<std::string> outputFiles;
+  bool hadError = false;
+  unsigned i = 0;
+  InputsAndOutputs.forEachInputProducingOutput(
+      [&](const InputFile &input) -> void {
+
+        StringRef outputArg = i < OutputFileArguments->size()
+                                  ? StringRef((*OutputFileArguments)[i])
+                                  : StringRef();
+
+        ++i;
+        Optional<std::string> outputFile = computeOutputFile(outputArg, input);
+        if (!outputFile) {
+          hadError = true;
+          return;
+        }
+        outputFiles.push_back(*outputFile);
+      });
+  return hadError ? None : Optional<std::vector<std::string>>(outputFiles);
+}
+
+Optional<std::string>
+OutputFilesComputer::computeOutputFile(StringRef outputArg,
+                                       const InputFile &input) const {
+
+  return outputArg.empty() ? deriveOutputFileFromInput(input)
+                           : !OutputDirectoryArgument.empty()
+                                 ? deriveOutputFileForDirectory(input)
+                                 : outputArg.str();
```

Comment:
I've replaced the expression ifs with imperative ifs. (I like expression-ifs better when the salient conditional computation is more functional than imperative, but that's just me.)
I did not invert the first predicate because that would have created a longer first alternative followed by a shorter second one. I believe that putting the shorter case first decreases the mental stack depth required to read the code, so I tend in that direction. I'll change it if you want.

Reply:
Hm, right. I forgot that OutputDirectoryArgument being non-empty is a sub-case of outputArg being non-empty. Maybe it would make sense to just check OutputDirectoryArgument first, since it's the most specific case?

Reply:
Hmm.... what should happen if the arguments are
    -primary-file p1.swift -primary-file p2.swift -o /tmp
As it stands I think the first output would be `/tmp/p1.<something>` and the second output would be `p2.<something>`.  So, now I'm convinced to try it the other way, as you suggest. I suspect none of the existing tests for non-batch will notice.

Reply:
OK, should be as we agreed now.